{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2015-07-14 11:55:06--  http://d28rh4a8wq0iu5.cloudfront.net/ads1/data/phix.fa\n",
      "Resolving d28rh4a8wq0iu5.cloudfront.net... 54.230.53.87, 54.192.54.164, 54.192.54.222, ...\n",
      "Connecting to d28rh4a8wq0iu5.cloudfront.net|54.230.53.87|:80... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 5528 (5.4K) [application/octet-stream]\n",
      "Saving to: 'phix.fa'\n",
      "\n",
      "phix.fa             100%[=====================>]   5.40K  --.-KB/s   in 0.003s \n",
      "\n",
      "2015-07-14 11:55:07 (1.60 MB/s) - 'phix.fa' saved [5528/5528]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!wget http://d28rh4a8wq0iu5.cloudfront.net/ads1/data/phix.fa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def readGenome(filename):\n",
    "    genome = ''\n",
    "    with open(filename, 'r') as f:\n",
    "        for line in f:\n",
    "            # ignore header line with genome information\n",
    "            if not line[0] == '>':\n",
    "                genome += line.rstrip()\n",
    "    return genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "genome = readGenome('phix.fa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def naive(p, t):\n",
    "    occurrences = []\n",
    "    for i in range(len(t) - len(p) + 1):\n",
    "        match = True\n",
    "        for j in range(len(p)):\n",
    "            if t[i+j] != p[j]:\n",
    "                match = False\n",
    "                break\n",
    "        if match:\n",
    "            occurrences.append(i)\n",
    "    return occurrences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[0, 5, 9]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "t = 'AGCTTAGATAGC'\n",
    "p = 'AG'\n",
    "naive(p, t)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import random\n",
    "def generateReads(genome, numReads, readLen):\n",
    "    ''' Generate reads from random positions in the given genome. '''\n",
    "    reads = []\n",
    "    for _ in range(numReads):\n",
    "        start = random.randint(0, len(genome)-readLen) - 1\n",
    "        reads.append(genome[start : start+readLen])\n",
    "    return reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "100 / 100 reads matched the genome exactly!\n"
     ]
    }
   ],
   "source": [
    "# Generate 100 reads of length 100\n",
    "reads = generateReads(genome, 100, 100)\n",
    "\n",
    "# Count how many reads match the genome exactly\n",
    "numMatched = 0\n",
    "for r in reads:\n",
    "    matches = naive(r, genome)\n",
    "    if len(matches) > 0:\n",
    "        numMatched += 1\n",
    "print('%d / %d reads matched the genome exactly!' % (numMatched, len(reads)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2018-03-21 12:33:44--  http://d28rh4a8wq0iu5.cloudfront.net/ads1/data/ERR266411_1.first1000.fastq\n",
      "Resolving d28rh4a8wq0iu5.cloudfront.net (d28rh4a8wq0iu5.cloudfront.net)... 52.84.206.23, 52.84.206.80, 52.84.206.233, ...\n",
      "Connecting to d28rh4a8wq0iu5.cloudfront.net (d28rh4a8wq0iu5.cloudfront.net)|52.84.206.23|:80... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 254384 (248K) [audio/mpeg]\n",
      "Saving to: ‘ERR266411_1.first1000.fastq’\n",
      "\n",
      "ERR266411_1.first10 100%[===================>] 248.42K   162KB/s    in 1.5s    \n",
      "\n",
      "2018-03-21 12:33:46 (162 KB/s) - ‘ERR266411_1.first1000.fastq’ saved [254384/254384]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!wget http://d28rh4a8wq0iu5.cloudfront.net/ads1/data/ERR266411_1.first1000.fastq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def readFastq(filename):\n",
    "    sequences = []\n",
    "    with open(filename) as fh:\n",
    "        while True:\n",
    "            fh.readline() # skip name line\n",
    "            seq = fh.readline().rstrip() # read base sequence\n",
    "            fh.readline() # skip placeholder line\n",
    "            fh.readline() # skip base quality line\n",
    "            if len(seq) == 0:\n",
    "                break\n",
    "            sequences.append(seq)\n",
    "    return sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Counter({'A': 28426, 'C': 21890, 'G': 19147, 'N': 6, 'T': 30531})"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import collections\n",
    "phix_reads = readFastq('ERR266411_1.first1000.fastq')\n",
    "count = collections.Counter()\n",
    "for read in phix_reads:\n",
    "    count.update(read)\n",
    "count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7 / 1000 reads matched the genome exactly!\n"
     ]
    }
   ],
   "source": [
    "numMatched = 0\n",
    "n = 0\n",
    "for r in phix_reads:\n",
    "    matches = naive(r, genome)\n",
    "    n += 1\n",
    "    if len(matches) > 0:\n",
    "        numMatched += 1\n",
    "print('%d / %d reads matched the genome exactly!' % (numMatched, n))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "459 / 1000 reads matched the genome exactly!\n"
     ]
    }
   ],
   "source": [
    "# Now let's try matching just the first 30 bases of each read\n",
    "numMatched = 0\n",
    "n = 0\n",
    "for r in phix_reads:\n",
    "    r = r[:30]  # just taking the first 30 bases\n",
    "    matches = naive(r, genome)\n",
    "    n += 1\n",
    "    if len(matches) > 0:\n",
    "        numMatched += 1\n",
    "print('%d / %d reads matched the genome exactly!' % (numMatched, n))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def reverseComplement(s):\n",
    "    complement = {'A': 'T', 'C': 'G', 'G': 'C', 'T': 'A', 'N': 'N'}\n",
    "    t = ''\n",
    "    for base in s:\n",
    "        t = complement[base] + t\n",
    "    return t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "932 / 1000 reads matched the genome exactly!\n"
     ]
    }
   ],
   "source": [
    "numMatched = 0\n",
    "n = 0\n",
    "for r in phix_reads:\n",
    "    r = r[:30]  # just taking the first 30 bases\n",
    "    matches = naive(r, genome)\n",
    "    matches.extend(naive(reverseComplement(r), genome))\n",
    "    n += 1\n",
    "    if len(matches) > 0:\n",
    "        numMatched += 1\n",
    "print('%d / %d reads matched the genome exactly!' % (numMatched, n))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'AACGTT'"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reverseComplement('AACGTT')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## week 1 programming assignment"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, implement a version of the naive exact matching algorithm that is strand-aware. That is, instead of looking only for occurrences of P in T, additionally look for occurrences of the reverse complement of P in T. If P is ACT, your function should find occurrences of both ACTand its reverse complement AGT in T.\n",
    "\n",
    "If P and its reverse complement are identical (e.g. AACGTT), then a given match offset should be reported only once. So if your new function is called naive_with_rc, then the old naive function and your new naive_with_rc function should return the same results when P equals its reverse complement."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def naive_with_rc(p, t):\n",
    "    occurrences = []\n",
    "    rc = reverseComplement(p)\n",
    "    for i in range(len(t) - len(p) + 1):\n",
    "        match = True\n",
    "        for j in range(len(p)):\n",
    "            if t[i+j] != p[j]:\n",
    "                match = False\n",
    "                break\n",
    "        if rc != p and not match:\n",
    "            match = True\n",
    "            for j in range(len(rc)):\n",
    "                if t[i+j] != rc[j]:\n",
    "                    match = False\n",
    "                    break\n",
    "        if match:\n",
    "            occurrences.append(i)\n",
    "    return occurrences"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "validations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[10, 23]\n",
      "[10, 24]\n",
      "offset of leftmost occurrence: 62\n",
      "# occurrences: 60\n"
     ]
    }
   ],
   "source": [
    "p = 'CCC'\n",
    "ten_as = 'AAAAAAAAAA'\n",
    "t = ten_as + 'CCC' + ten_as + 'GGG' + ten_as\n",
    "occurrences = naive_with_rc(p, t)\n",
    "print(occurrences)\n",
    "\n",
    "p = 'CGCG'\n",
    "t = ten_as + 'CGCG' + ten_as + 'CGCG' + ten_as\n",
    "occurrences = naive_with_rc(p, t)\n",
    "print(occurrences)\n",
    "\n",
    "phix_genome = readGenome('phix.fa')\n",
    "occurrences = naive_with_rc('ATTA', phix_genome)\n",
    "print('offset of leftmost occurrence: %d' % min(occurrences))\n",
    "print('# occurrences: %d' % len(occurrences))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, download and parse the lambda virus genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "lvgenome = readGenome('lambda_virus.fa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "matches = naive_with_rc('AGTCGA', lvgenome)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pattern AGGT and its RC matched the lambda virus genome 306 times\n"
     ]
    }
   ],
   "source": [
    "print('pattern %s and its RC matched the lambda virus genome %d times' % ('AGGT', len(matches)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pattern TTAA and its RC matched the lambda virus genome 195 times\n"
     ]
    }
   ],
   "source": [
    "print('pattern %s and its RC matched the lambda virus genome %d times' % ('TTAA', len(matches)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pattern ACTAAGT offset of leftmost occurrence: 26028\n"
     ]
    }
   ],
   "source": [
    "print('pattern %s offset of leftmost occurrence: %d' % ('ACTAAGT', min(matches)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pattern AGTCGA offset of leftmost occurrence: 450\n"
     ]
    }
   ],
   "source": [
    "print('pattern %s offset of leftmost occurrence: %d' % ('AGTCGA', min(matches)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Implement the naive_2mm function\n",
    "def naive_2mm(p, t):\n",
    "    occurrences = []\n",
    "    for i in range(len(t) - len(p) + 1):\n",
    "        match = True\n",
    "        mismatches = 0\n",
    "        for j in range(len(p)):\n",
    "            if t[i+j] != p[j] and mismatches == 2:\n",
    "                match = False\n",
    "                break\n",
    "            elif t[i+j] != p[j]: \n",
    "                mismatches += 1\n",
    "        if match:\n",
    "            occurrences.append(i)\n",
    "    return occurrences"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "validations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[10, 24, 38]\n",
      "offset of leftmost occurrence: 10\n",
      "# occurrences: 79\n"
     ]
    }
   ],
   "source": [
    "p = 'CTGT'\n",
    "ten_as = 'AAAAAAAAAA'\n",
    "t = ten_as + 'CTGT' + ten_as + 'CTTT' + ten_as + 'CGGG' + ten_as\n",
    "occurrences = naive_2mm(p, t)\n",
    "print(occurrences)\n",
    "\n",
    "phix_genome = readGenome('phix.fa')\n",
    "occurrences = naive_2mm('GATTACA', phix_genome)\n",
    "print('offset of leftmost occurrence: %d' % min(occurrences))\n",
    "print('# occurrences: %d' % len(occurrences))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "matches = naive_2mm('AGGAGGTT', lvgenome)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pattern TTCAAGCC matched the lambda virus genome 191 times\n"
     ]
    }
   ],
   "source": [
    "print('pattern %s matched the lambda virus genome %d times' % ('TTCAAGCC', len(matches)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pattern AGGAGGTT offset of leftmost occurrence: 49\n"
     ]
    }
   ],
   "source": [
    "print('pattern %s offset of leftmost occurrence: %d' % ('AGGAGGTT', min(matches)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2018-03-21 17:38:38--  https://d28rh4a8wq0iu5.cloudfront.net/ads1/data/ERR037900_1.first1000.fastq\n",
      "Resolving d28rh4a8wq0iu5.cloudfront.net (d28rh4a8wq0iu5.cloudfront.net)... 52.84.206.145, 52.84.206.80, 52.84.206.227, ...\n",
      "Connecting to d28rh4a8wq0iu5.cloudfront.net (d28rh4a8wq0iu5.cloudfront.net)|52.84.206.145|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 241626 (236K) [application/octet-stream]\n",
      "Saving to: ‘ERR037900_1.first1000.fastq’\n",
      "\n",
      "ERR037900_1.first10 100%[===================>] 235.96K   167KB/s    in 1.4s    \n",
      "\n",
      "2018-03-21 17:38:41 (167 KB/s) - ‘ERR037900_1.first1000.fastq’ saved [241626/241626]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!wget https://d28rh4a8wq0iu5.cloudfront.net/ads1/data/ERR037900_1.first1000.fastq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def readFastq(filename):\n",
    "    sequences = []\n",
    "    qualities = []\n",
    "    with open(filename) as fh:\n",
    "        while True:\n",
    "            fh.readline() # skip name line\n",
    "            seq = fh.readline().rstrip() # read base sequence\n",
    "            fh.readline() # skip placeholder line\n",
    "            qual = fh.readline().rstrip() #base quality line\n",
    "            if len(seq) == 0:\n",
    "                break\n",
    "            sequences.append(seq)\n",
    "            qualities.append(qual)\n",
    "    return sequences, qualities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Counter({'A': 24057, 'C': 29665, 'G': 22888, 'N': 914, 'T': 22476})"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import collections\n",
    "phix2_reads, phix2_qualities = readFastq('ERR037900_1.first1000.fastq')\n",
    "count = collections.Counter()\n",
    "for read in phix2_reads:\n",
    "    count.update(read)\n",
    "count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def phred33ToQ(qual):\n",
    "    return ord(qual) - 33"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phred33ToQ('#')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0, 0, 17723, 0, 2, 11, 11, 28, 23, 55, 100, 111, 86, 174, 185, 272, 317, 259, 390, 1523, 2782, 762, 286, 413, 403, 538, 351, 694, 971, 777, 1024, 1449, 1341, 1312, 1916, 2233, 3025, 4043, 6640, 45696, 2074, 0, 0, 0, 0, 0, 0, 0, 0, 0]\n"
     ]
    }
   ],
   "source": [
    "def createHist(qualities):\n",
    "    # Create a histogram of quality scores\n",
    "    hist = [0]*50\n",
    "    for qual in qualities:\n",
    "        for phred in qual:\n",
    "            q = phred33ToQ(phred)\n",
    "            hist[q] += 1\n",
    "    return hist\n",
    "h = createHist(phix2_qualities)\n",
    "print(h)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXcAAAD8CAYAAACMwORRAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAHsNJREFUeJzt3XmUnGWZ9/Hv1dVVSTr70llI0kmABBJCwtJAEAEBkQTF\niOIMDIvjxnBGGPV1RlFndEZ03mHQGT1HFBlEZsABHYwSYgQRx8ALggmQfSOELJ3O1oHO0llqu94/\nqqqpVLqTSlKdqr7r9zknJ/089SR13SH9485V93M/5u6IiEhYaspdgIiIlJ7CXUQkQAp3EZEAKdxF\nRAKkcBcRCZDCXUQkQAp3EZEAKdxFRAKkcBcRCVBtud54yJAhPnbs2HK9vYhIt/TKK6+0uHv9ka4r\nW7iPHTuWBQsWlOvtRUS6JTNbX8x1asuIiARI4S4iEiCFu4hIgBTuIiIBUriLiARI4S4iEiCFu4hI\ngBTuIoFq3RvnyUXN5S5DykThLhKo2YuauePR13i7LV7uUqQMFO4igdqfSAFwIJkucyVSDgp3kUAl\nUp79WeFejRTuIoHKhXpc4V6VFO4igcqFezI7g5fqonAXCZTaMtVN4S4SqHhSbZlqpnAXCVRuxp7Q\napmqpHAXCVSyvS2jnns1UriLBKp95p7WzL0aKdxFAhVXW6aqKdxFAtU+c1dbpiop3EUCldRSyKqm\ncBcJVHtbRuFelRTuIoFSW6a6KdxFAqU7VKubwl0kUEm1Zaqawl0kUPHszF3bD1QnhbtIoLQrZHVT\nuIsEKqG2TFVTuIsEKqm2TFVTuIsE6p3tB9SWqUZFhbuZTTezVWa2xszu7OD1/mb2pJktMrNlZvbx\n0pcqIkdDbZnqdsRwN7MIcC8wA5gE3GBmkwou+wyw3N2nAu8BvmNmsRLXKiJHIbdhWFK7QlalYmbu\n5wNr3H2tu8eBx4CZBdc40NfMDOgDvAUkS1qpiByV3E1McbVlqlIx4T4S2Jh33JQ9l+/7wESgGVgC\nfNbdNV0QKRN3b9/HXW2Z6lSqD1SvAhYCJwFnAd83s36FF5nZrWa2wMwWbN++vURvLSKFUmnHsxN2\nhXt1KibcNwGj845HZc/l+zgwyzPWAG8Cpxf+Ru5+v7s3untjfX39sdYsIkeQv1mYwr06FRPu84Hx\nZjYu+yHp9cDsgms2AFcAmNkw4DRgbSkLFZHi5a9t166Q1an2SBe4e9LMbgeeBiLAg+6+zMxuy75+\nH3AX8JCZLQEM+JK7t3Rh3SJyGMmDwl0z92p0xHAHcPe5wNyCc/flfd0MvK+0pYnIsVJbRnSHqkiA\n8gM9rrZMVVK4iwQov+ee1My9KincRQKUVFum6incRQKUC/Re0YhWy1QphbtIgHJtmd49IsSTmrlX\nI4W7SIBym4bVxWrVlqlSCneRACXTmVZMXSzS/rVUF4W7SIBybZlesUj7LF6qi8JdJEC5QO8dq9Vj\n9qqUwl0kQLkVMnWxiHruVUrhLhKgXKDXxSKkPbMFsFQXhbtIgNrDvUftQcdSPRTuIgFqb8tEI9lj\nhXu1UbiLBOjQmbvaMtVG4S4SoFy4945p5l6tFO4iAcpfLQNoC4IqpHAXCdA7q2UybRndpVp9FO4i\nAUqk0tQY9IjWtB9LdVG4iwQonkoTjdQQjWS+xdWWqT4Kd5EAJVNONFJDLKKZe7VSuIsEKJFKE41Y\n+8xdSyGrj8JdJECJ9raMAXqOaiXZG0+ekDaZwl0kQPFkpi1Tm+u5K9wrwh9WbeN9//4c//H82i5/\nr9oufwcROeGS6UxbJqa2TEXYvvsA35iznCcXNXNKfW8axwzs8vdUuIsEqL0tU2vtx3Livd0W58EX\n3uShF9ZxIJnm8++dwG3vOZketZEuf2+Fu0iAcm2ZqFbLlMWabbv5nwVNPPLSetriKWZMHs7fXnUa\np9T3OWE1KNxFApRIpYnW1hCtUVvmRGlu3ceTi5qZvaiZZc27qDF4/5STuP2yUzlteN8TXo/CXSRA\niVSaaI2pLdPF0mnn5ws2MuvVTfxp3VsAnDV6AF/7wCQ+MGUEQ/v1LFttCneRAOVuYlJbpmv9fMFG\n7py1hPFD+/C375vANVNPYszg3uUuC1C4iwQpnkrTLxbV9gNdKJ127n9uLWeO7M/s2y/CzMpd0kG0\nzl0kQIlUmljeUkjtCll6z6zYytqWNv7q0pMrLthB4S4SpEQqTW1NDbXZO1QTmrmXlLtz37w3aBhU\nx/Qzhpe7nA4p3EUClEw50doaamv0gWpXmL/ubV7b0MqnLx7XfhdwpanMqkTkuMSzG4eZZVozcS2F\nLKkfzXuDQb1jXHfu6HKX0qmiwt3MppvZKjNbY2Z3dnLNe8xsoZktM7N5pS1TRI5Gpuee+faORkwz\n9xJauWUXz67cxscuHEuvWNffaXqsjrhaxswiwL3AlUATMN/MZrv78rxrBgA/AKa7+wYzG9pVBYvI\nkSVS3t5vj9bWaFfIEkmnnb//5VL694pyy4Vjyl3OYRUzcz8fWOPua909DjwGzCy45i+AWe6+AcDd\nt5W2TBE5Grm9ZQBqa9SWKZWHX1rPgvVv8w8fmMTA3rFyl3NYxYT7SGBj3nFT9ly+CcBAM/uDmb1i\nZreUqkAROXr5bZmY2jIl0fT2Xu5+aiWXTKjnI+cURmDlKdVNTLXAucAVQC/gj2b2kruvzr/IzG4F\nbgVoaGgo0VuLSKFE9g5VyLRlFO7Hx935yi+XAvDP106uyHXthYqZuW8C8j8SHpU9l68JeNrd29y9\nBXgOmFr4G7n7/e7e6O6N9fX1x1qziBxGOu2k0nk994jC/Xhs272fLz6+mOdWb+dL009n1MC6cpdU\nlGJm7vOB8WY2jkyoX0+mx57vCeD7ZlYLxIALgH8vZaEiUpxEOhPk7TP3SI12hTwG++Ipfvz/1vLD\nP7zBgWSaT188jpunVfaHqPmOGO7unjSz24GngQjwoLsvM7Pbsq/f5+4rzOwpYDGQBh5w96VdWbiI\ndCwX5FoKeWz2HEjyyEvreeD5tbTsiXPVGcO4c8ZExg2pjA3BilVUz93d5wJzC87dV3B8D3BP6UoT\nkWOR22ogqrZM0fbFU7y0dgfzVm/nVws30bo3wcXjh/DZK8bTOHZQucs7JtoVUiQwuSCvzZ+5J9WW\n6czDL63nrjnLiSfT9IzWcOmEem679BTObuj655x2JYW7SGAS6cK2TA27E8lyllSxEqk03/vdaiaN\n6McX3jeB88YOome0cu86PRraW0YkMO1tmexTmGKRGpJptWU6Mm/Vdlr2xLn9slO5eHx9MMEOCneR\n4OTaMu13qKot06nHX2liSJ8Yl54W3tJshbtIYOK5nntN/lJIzdwLvdUW59mVW/nQWSPb/0cYkvBG\nJFLl2pdC5rVl4gr3QzyxcBOJlHNd46hyl9IlFO4igUmmDr2JKambmA7x+CtNnDmyP6cP71fuUrqE\nwl0kMPHCcK/VTUyFljfvYlnzLq47N8xZOyjcRYKTa8scvOWvwj3fzxdsJBap4YNTTyp3KV1G4S4S\nmMI7VGPaFfIgK7fs4qcvr+eDZ51U8XuyHw+Fu0hgkodsHGbaOCwrmUrzpccX07dnlK9cPbHc5XQp\nhbtIYOIFbZlopIZU2kmnFfA/eWEdi5p28o8fPINBAc/aQeEuEpxcWyZ/+wF4ZyvgarWupY3vPLOK\n904cyjVTRpS7nC6ncBcJzDsbh+V2hbTs+eqdue+NJ/n8zxcSranhmx86s1s8Sel4KdxFApPbOCxa\nOHNPVufMfX8ixV89/AqLNrbyr9dNYXj/nuUu6YTQrpAigem0LVOFK2YSqTS3//drPP96C/dcN4UZ\nZ4bfjsnRzF0kMO0bh+VtPwDvzOirRTyZ5nM/W8jvVmzlGzPP4KONo4/8iwKimbtIYBIFG4fleu/V\n1JZpO5Dktkde4fnXW/jq1RO55cKx5S7phFO4iwTmnTtU33nMXuZ8dYT7jj0H+MRD81myaSd3f+RM\n/vy8hnKXVBYKd5HAJFJpohFrXxGSC/dq2IJg1Zbd3PbIKzS37uNHNzdy5aRh5S6pbBTuIoHJhPs7\nH6fltv4NfSnkL19r4iuzltK7Ry0//dQF3fbB1qWicBcJTCLl1Na8s447F/TJgGfu//TkMn7ywjrO\nHzeI799wNkP7Vcdyx8NRuIsEJp5KE6t9Z+ae+2A11LbM61t385MX1nHD+aO5a+ZkagN8qtKx0J+C\nSGCSVdaWeXLxZmoMPn/lBAV7Hv1JiAQmkfKDwj3kO1TdnTmLmrlg3GCG9lUrJp/CXSQw8VS6fW07\n5PXcA9w4bPnmXaxtaeMDU6vnztNiKdxFApNIptvvSoX8pZDhtWXmLN5MpMaYMVnhXkjhLhKYZLqw\nLRPmHaruzpzFzbzrlMHB781+LBTuIoHJ3cSUE+odqoubdrLxrX1cE/BzUI+Hwl0kMPFk+qBVI6GG\n+5OLmolGjKsmDS93KRVJ4S4SmETq4J57+66QAfXc02nn10s2c8n4evrXRctdTkVSuIsEJtNzz2vL\ntK9zD2fm/vBL69m8cz8fOXdUuUupWAp3kcDEkwffxJS7QzWUcG96ey93P7WSSybUM2OyWjKdUbiL\nBKZw47DcLD6EpZDuzld+uRSAf752clU8C/VYFRXuZjbdzFaZ2Rozu/Mw151nZkkzu650JYrI0cjc\nofpO6JkZ0YgFMXP/xaubeG71dr40/XRGDawrdzkV7YjhbmYR4F5gBjAJuMHMJnVy3d3Ab0tdpIgU\nr3BvGcismOnuu0IuadrJXXOW0zhmIDdPG1PucipeMTP384E17r7W3ePAY8DMDq67A/gFsK2E9YnI\nUYqnnGjtoeHeXVfLuDv//fIGPvLDF+kdi3DPR6dSU6N2zJEUs+XvSGBj3nETcEH+BWY2ErgWuAw4\nr2TVichRS6TSRAvCLxqxbrnl79Zd+7n7qZXMenUTl0yo57t/fpbuRi1SqfZz/y7wJXdPH+4DDjO7\nFbgVoKGhOp9rKNLVCj9QhezMvZtsP7A/keK1Da088vJ6nl66hZQ7n3vveO64fDwRzdiLVky4bwJG\n5x2Pyp7L1wg8lg32IcDVZpZ091/lX+Tu9wP3AzQ2NnbPfyOKVLhEKt1JW6Zyw33Ntt187YllrN3e\nxpZd+wHo3yvKxy8ay03TxjBmcO8yV9j9FBPu84HxZjaOTKhfD/xF/gXuPi73tZk9BMwpDHYR6Xru\nfsh+7pBpyyTSlTufeviP61mw/m2umXISYwfXccrQPlx22lB6xSLlLq3bOmK4u3vSzG4HngYiwIPu\nvszMbsu+fl8X1ygiRUpmA/zQnnvltmXSaec3S7dw2Wn1fOfPppa7nGAU1XN397nA3IJzHYa6u//l\n8ZclIsci13rpTm2Z1za+zbbdB7Qne4npDlWRgCSS2Zl7R22ZCl0K+ZslW4hGjMsnDi13KUFRuIsE\nJJF9lF4scmhbphKXQrpnWjLvPnUI/Xpqd8dSUriLBKS9LVMwc4/VVuYdqks27WRT6z5mnKmWTKkp\n3EUCkmvL1BaEe21NZbZlfrN0C5Ea48qJw8pdSnAU7iIBibfP3DtYLVNhM3d356mlW7jw5MEM1F2n\nJadwFwlIsr3nXvCBam3l9dxXbd3Nmy1tTNee7F1C4S4SkM5Wy8QiNSQrqC2TTjvffno1kRrjfWeo\nJdMVFO4iAcnNzmsPactU1n7udz+9kt+t2Mo/vH8iQ/v2LHc5QVK4iwQkF+CFbZnaCuq5/8+Cjfxo\n3lpumtbAx941ttzlBEvhLhKQXOul8A7VWKSGeAVsPzBv9Xa+8sslvPvUIXz9mjP0mLwuVKotf0Wk\nAnS2zr3cd6im084P573Bd367ignD+nLvjeccUqOUlsJdJCDtPfcONg7LraQ50bbu2s+XZy3h9yu3\n8cGpJ/F/P3wmvXsoerqa/oRFAtLec+/kMXvu3mWtkJ37EmzYsZfWfXFa9yZYvnkXz63ezrLmXUQj\nxjdmnsHN08aoFXOCKNxFAnK4tkzmdSdWW5pwTaWdxU2tPLe6hXmrt7FwYyv5W8bX1hjnjBnI3111\nGtMnD+eU+j4leV8pjsJdJCC5vnpHd6hmXk8fMqvP5+54B635tDtbdx9gfUsba1vaePnNt3j+9e20\n7k1gBlNG9uf2y07ljJH9GVgXY0BdlJMG9KKP2i9loz95kYB0PnOvOej1Qu7Oz+Zv5FtzV7B7f/KI\n7zOkTw8uP30ol06o5+Lx9XpodQVSuIsEJPe0pUPCvTYX7odOy3fsOcCds5bwzPKtTDt5EBeePKTD\n33tovx6MGVRHw+A6Turfixo9rLqiKdxFAtJZWybW3nNPszeeZFnzLhZtbGXhxlZeWNNCWzzF379/\nIp+4aJxCOxAKd5GA5B7WUThzr63JHN/4wMtseGsvqewnnyMH9OJdpwzhjitO5fTh/U5ssdKlFO4i\nAYl30pY5bXhfGgbVMXpQHddMGcGUUQOYMrq/9nUJmMJdJCBbdu5ncO8YkYLWyuSR/Xnui5eVqSop\nB93/KxKQdTvaGDukd7nLkAqgcBcJyLqWvYwZXFfuMqQCKNxFArEvnmLLrv2MG6yZuyjcRYKx/q02\nAMaoLSMo3EWCsa4lE+6auQso3EWCsW7HXgDGDFHPXRTuIsFY19LG4N4x+vWMlrsUqQAKd5FAaBmk\n5FO4iwRCyyAln8JdJABaBimFFO4iAdAySCmkcBcJgJZBSiGFu0gAtAxSChUV7mY23cxWmdkaM7uz\ng9dvNLPFZrbEzF40s6mlL1VEOqNlkFLoiOFuZhHgXmAGMAm4wcwmFVz2JnCpu58J3AXcX+pCRaRz\nb7ZoGaQcrJiZ+/nAGndf6+5x4DFgZv4F7v6iu7+dPXwJGFXaMkXkcNbv0DJIOVgx4T4S2Jh33JQ9\n15lPAr85nqJEpHhaBikdKemTmMzsMjLh/u5OXr8VuBWgoaGhlG8tUrW0DFI6UszMfRMwOu94VPbc\nQcxsCvAAMNPdd3T0G7n7/e7e6O6N9fX1x1KviBTQMkjpSDHhPh8Yb2bjzCwGXA/Mzr/AzBqAWcDN\n7r669GWKSGe0DFI6csS2jLsnzex24GkgAjzo7svM7Lbs6/cBXwMGAz8wM4Ckuzd2XdkiAuDu/O/K\nbYwc0EvLIOUgRfXc3X0uMLfg3H15X38K+FRpSxORI3nxjR28/OZbfP2awtXJUu10h6pIN+Xu/Nsz\nqxnRvyc3nK8FCnIwhbtINzVv9XZeWf82n7nsVHpGI+UuRyqMwl2kG8rN2kcO6MWfNY4+8i+QqqNw\nF+lm4sk0j83fyOKmnfzNFacSq9W3sRyqpDcxiUjXcHd+vWQzv3qtmT++0UJbPMWEYX348Dna6UM6\npnAXqXCvrH+Lu+asYOHGVkYO6MW154zk0glDuejUwUQjmrVLxxTuIhVm5ZZdPLtiG+t3tLFm2x5e\n3dDK0L49+NfrpvCRc0YRqbFylyjdgMJdpILsi6f46H1/ZPf+JPV9ezB2cB1fuHICn7x4HHUxfbtK\n8fS3RaSCzFnczO79SR799DQuPGVwucuRbkwNO5EK8uifNnBKfW+mnTyo3KVIN6dwF6kQq7bs5tUN\nrdxwfgPZPZpEjpnCXaRCPPqnDcQiNVreKCWhcBepAPsTKWa92sRVk4czqHes3OVIABTuIhVg7pLN\n7Nqf5IbztZWAlIbCXaTMdu1P8NCL6xg7uI4LT9YKGSkNLYUUKRN35+llW/j67GVs232Ae66bqg9S\npWQU7iIn2Oad+3hu9XbmLN7M86+3MHFEP+6/uZGpoweUuzQJiMJdpAul086KLbtYuLGVRRtbeW1D\nK69v2wPAsH49+MrVp/OJi8ZRqz1ipMQU7iIllk5ndnB8dsVWnn+9hR1tcQAG9Y4xZVR/Pto4iksn\nDGXCsD5qw0iXUbiLlNiDL7zJN3+9gsG9Y1wyoZ6Lxw/hvLGDGDWwl8JcThiFu0gJrd/Rxrd/u4rL\nTx/KA7c0UqMdHKVM1OgTKRF3585fLCFaU8O3rp2sYJeyUriLlMjP5m/kj2t38OWrJzKif69ylyNV\nTuEuUgKLm1r51q9XMO3kQVx/nu4ylfJTz12kSO7OEwub+d2KrZw7ZiCXTqinVyzCPU+vYtarmxjS\nJ8a/fHiK2jFSERTuIkVo3Rvnq79ayq8Xb6Z/ryhzFm8GoMagNlLDbZeewl9fdgr9ekbLXKlIhsJd\npAObd+7j9yu30bo3wc59CWYvbKZlzwH+7qrTuO3SU2hu3ce81dvZ+PZebrpgDKMH1ZW7ZJGDKNxF\nCmzYsZeP/uhFtu46AEDPaA3jh/blP25p5MxR/QEYPaiOm6aNKWeZIoelcBfJs3nnPm788UvsT6SZ\n9dfvYtKIfvSMRspdlshRU7iLZG3ffYAbH3iZt9sS/PRTF2gjL+nWFO5StdydvfEUL63dwRMLm3lm\n+VYc578+oWCX7k/hLsHbF0+x/q02lm3axaKmVhY17aS5dR879yaIp9IADKiLcu05I7nxggbOOKl/\nmSsWOX4KdwmKu7Nyy27mrd7OC2taWL11d/sHowC9YxHOHNWfK04fyoC6GAPqopw2rC8XnTqEWK3u\n6ZNwKNyl23N3lmzayeyFzcxZvJktu/YDcPrwvlw8vp6xg+toGNybicP7cnJ9HyK6yUiqQFHhbmbT\nge8BEeABd/+Xgtct+/rVwF7gL9391RLXKlVi9/4ESzbtZHHTTtZs20Pa/ZBr0mln9/4krfsSbNm5\nn02t+4hGjEsnDOX/XDmBSybUM7x/zzJUL1IZjhjuZhYB7gWuBJqA+WY2292X5102Axif/XEB8MPs\nzyKHlUyl2bU/ydrte3hu9Xbmrd7O4k07yeX5sH49iHbwlCIz6NczyoC6KGc3DOCOy09lxuQR9K/T\nHaIiUNzM/XxgjbuvBTCzx4CZQH64zwT+y90deMnMBpjZCHffXPKKpSIdSKbYuTdB674ErXsTrNm2\nh0UbW1nU1ErLngOd/Jo0u/cn249rDM4aPYA7Lh/P2Q0DmDpqAIN6x07UEESCUky4jwQ25h03ceis\nvKNrRgIlD/d5q7fzzTnLj3yhdDkH2g4kad2bYF8idcjrA+uiTBk1gHPGDKSjLnc0UsOAuij9e0UZ\n0b8nF548RDNvkRI5oR+omtmtwK0ADQ0Nx/R79OlRy/hhfUpZlhyH3rFaBtRFGVAXo1+vKAOzYd0w\nqI6GQXV6rJxImRQT7puA/A2qR2XPHe01uPv9wP0AjY2Nh35KVoRzxwzk3DHnHssvFRGpGsUs7J0P\njDezcWYWA64HZhdcMxu4xTKmATvVbxcRKZ8jztzdPWlmtwNPk1kK+aC7LzOz27Kv3wfMJbMMcg2Z\npZAf77qSRUTkSIrqubv7XDIBnn/uvryvHfhMaUsTEZFjpfutRUQCpHAXEQmQwl1EJEAKdxGRACnc\nRUQCZN7Bjnsn5I3NtgPrj/GXDwFaSlhOd1GN467GMUN1jrsaxwxHP+4x7l5/pIvKFu7Hw8wWuHtj\nues40apx3NU4ZqjOcVfjmKHrxq22jIhIgBTuIiIB6q7hfn+5CyiTahx3NY4ZqnPc1Thm6KJxd8ue\nu4iIHF53nbmLiMhhdLtwN7PpZrbKzNaY2Z3lrqcrmNloM/tfM1tuZsvM7LPZ84PM7Bkzez3788By\n11pqZhYxs9fMbE72uBrGPMDMHjezlWa2wswurJJxfz7793upmT1qZj1DG7eZPWhm28xsad65Tsdo\nZl/OZtsqM7vqeN67W4V73sO6ZwCTgBvMbFJ5q+oSSeAL7j4JmAZ8JjvOO4Fn3X088Gz2ODSfBVbk\nHVfDmL8HPOXupwNTyYw/6HGb2Ujgb4BGd59MZjvx6wlv3A8B0wvOdTjG7Pf49cAZ2V/zg2zmHZNu\nFe7kPazb3eNA7mHdQXH3ze7+avbr3WS+2UeSGet/Zi/7T+BD5amwa5jZKOD9wAN5p0Mfc3/gEuDH\nAO4ed/dWAh93Vi3Qy8xqgTqgmcDG7e7PAW8VnO5sjDOBx9z9gLu/Seb5GOcf63t3t3Dv7EHcwTKz\nscDZwMvAsLwnXG0BhpWprK7yXeCLQDrvXOhjHgdsB36SbUc9YGa9CXzc7r4J+DawAdhM5ultvyXw\ncWd1NsaS5lt3C/eqYmZ9gF8An3P3XfmvZR+QEsxSJzP7ALDN3V/p7JrQxpxVC5wD/NDdzwbaKGhF\nhDjubJ95Jpn/uZ0E9Dazm/KvCXHchbpyjN0t3It6EHcIzCxKJth/6u6zsqe3mtmI7OsjgG3lqq8L\nXAR80MzWkWm3XW5mjxD2mCEzO2ty95ezx4+TCfvQx/1e4E133+7uCWAW8C7CHzd0PsaS5lt3C/di\nHtbd7ZmZkenBrnD3f8t7aTbwsezXHwOeONG1dRV3/7K7j3L3sWT+u/7e3W8i4DEDuPsWYKOZnZY9\ndQWwnMDHTaYdM83M6rJ/368g89lS6OOGzsc4G7jezHqY2ThgPPCnY34Xd+9WP8g8iHs18Abw1XLX\n00VjfDeZf6otBhZmf1wNDCbz6frrwO+AQeWutYvG/x5gTvbr4McMnAUsyP73/hUwsErG/U/ASmAp\n8DDQI7RxA4+S+UwhQeZfaZ883BiBr2azbRUw43jeW3eoiogEqLu1ZUREpAgKdxGRACncRUQCpHAX\nEQmQwl1EJEAKdxGRACncRUQCpHAXEQnQ/wc1Ig+DJSKXMwAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x10cca6b50>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#function to fish the distributions of thread offsets where the hash (#) quality is reported\n",
    "#to answer question 7 of the programming assignment\n",
    "%matplotlib inline\n",
    "import matplotlib.pyplot as plt\n",
    "def findHashByPos(qualities):\n",
    "    ''' Find the # ratio at each position in the read '''\n",
    "    # Keep track of the number of G/C bases and the total number of bases at each position\n",
    "    hashqual = [0] * 100\n",
    "    totals = [0] * 100\n",
    "    for read in qualities:\n",
    "        for i in range(len(read)):\n",
    "            if read[i] == '#':\n",
    "                hashqual[i] += 1\n",
    "            totals[i] += 1\n",
    "    # Divide # counts by total counts to get the average at each position\n",
    "    for i in range(len(hashqual)):\n",
    "        if totals[i] > 0:\n",
    "            hashqual[i] /= float(totals[i])\n",
    "    return hashqual\n",
    "\n",
    "hashqual = findHashByPos(phix2_qualities)\n",
    "plt.plot(range(len(hashqual)), hashqual)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0: 0.0\n",
      "1: 0.0\n",
      "2: 0.0\n",
      "3: 0.0\n",
      "4: 0.0\n",
      "5: 0.0\n",
      "6: 0.0\n",
      "7: 0.0\n",
      "8: 0.0\n",
      "9: 0.0\n",
      "10: 0.0\n",
      "11: 0.0\n",
      "12: 0.0\n",
      "13: 0.0\n",
      "14: 0.0\n",
      "15: 0.0\n",
      "16: 0.0\n",
      "17: 0.0\n",
      "18: 0.0\n",
      "19: 0.0\n",
      "20: 0.0\n",
      "21: 0.0\n",
      "22: 0.0\n",
      "23: 0.0\n",
      "24: 0.0\n",
      "25: 0.0\n",
      "26: 0.0\n",
      "27: 0.0\n",
      "28: 0.0\n",
      "29: 0.0\n",
      "30: 0.0\n",
      "31: 0.0\n",
      "32: 0.0\n",
      "33: 0.001\n",
      "34: 0.002\n",
      "35: 0.005\n",
      "36: 0.005\n",
      "37: 0.005\n",
      "38: 0.006\n",
      "39: 0.01\n",
      "40: 0.013\n",
      "41: 0.013\n",
      "42: 0.013\n",
      "43: 0.014\n",
      "44: 0.019\n",
      "45: 0.023\n",
      "46: 0.025\n",
      "47: 0.028\n",
      "48: 0.029\n",
      "49: 0.034\n",
      "50: 0.049\n",
      "51: 0.05\n",
      "52: 0.06\n",
      "53: 0.056\n",
      "54: 0.064\n",
      "55: 0.086\n",
      "56: 0.09\n",
      "57: 0.093\n",
      "58: 0.097\n",
      "59: 0.106\n",
      "60: 0.142\n",
      "61: 0.144\n",
      "62: 0.152\n",
      "63: 0.156\n",
      "64: 0.184\n",
      "65: 0.284\n",
      "66: 0.919\n",
      "67: 0.287\n",
      "68: 0.29\n",
      "69: 0.293\n",
      "70: 0.303\n",
      "71: 0.303\n",
      "72: 0.303\n",
      "73: 0.305\n",
      "74: 0.309\n",
      "75: 0.313\n",
      "76: 0.314\n",
      "77: 0.318\n",
      "78: 0.329\n",
      "79: 0.346\n",
      "80: 0.39\n",
      "81: 0.392\n",
      "82: 0.398\n",
      "83: 0.409\n",
      "84: 0.431\n",
      "85: 0.473\n",
      "86: 0.484\n",
      "87: 0.497\n",
      "88: 0.512\n",
      "89: 0.534\n",
      "90: 0.572\n",
      "91: 0.576\n",
      "92: 0.583\n",
      "93: 0.591\n",
      "94: 0.613\n",
      "95: 0.644\n",
      "96: 0.649\n",
      "97: 0.654\n",
      "98: 0.661\n",
      "99: 0.67\n"
     ]
    }
   ],
   "source": [
    "for i in range(len(hashqual)):\n",
    "    print str(i)+\": \"+str(hashqual[i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
